File: src/product_factory/host_mcp/tools.py
"""MCP tool handlers — call HostService directly; return HostResponse JSON."""

from __future__ import annotations

import uuid
from decimal import Decimal
from pathlib import Path
from typing import Any

from product_factory.domain.budgets import RunBudget
from product_factory.domain.runs import ArtifactOverride, RunRequest
from product_factory.host.protocol import HostResponse
from product_factory.host.service import HostService
from product_factory.workflows.artifacts import ArtifactOverrideError, normalize_overrides

TOOL_NAMES = (
    "pf_submit",
    "pf_status",
    "pf_tail",
    "pf_inspect",
    "pf_approve",
    "pf_reject",
    "pf_cancel",
    "pf_export",
    "pf_materialize",
    "pf_materialize_all",
)

_WORKFLOW_VALUES = {
    "architecture",
    "technical_plan",
    "code_change",
    "repository_change",
    "repository_investigation",
    "quality_gate",
}


def _as_json(response: HostResponse) -> dict[str, Any]:
    return response.model_dump(mode="json")


def _failure(code: str, message: str, **kwargs: Any) -> dict[str, Any]:
    return _as_json(HostResponse.failure(code=code, message=message, **kwargs))


def tool_schemas() -> list[dict[str, Any]]:
    """JSON Schema definitions for MCP ``tools/list``."""
    return [
        {
            "name": "pf_submit",
            "description": (
                "Submit a curated Product Factory request. Returns HostResponse "
                "with run_id + subscription. Do not dump full chat transcripts — "
                "pass named request text only."
            ),
            "inputSchema": {
                "type": "object",
                "properties": {
                    "request_text": {
                        "type": "string",
                        "description": "Curated operator request text",
                    },
                    "workflow": {
                        "type": "string",
                        "description": (
                            "Workflow pack id: repository_investigation, "
                            "technical_plan, quality_gate, repository_change, "
                            "code_change, architecture"
                        ),
                        "default": "code_change",
                    },
                    "repository_path": {
                        "type": "string",
                        "description": "Absolute or cwd-relative repo path",
                    },
                    "budget_usd": {
                        "type": "number",
                        "description": "Max USD budget for the run",
                        "default": 3.0,
                    },
                    "validation_commands": {
                        "type": "array",
                        "items": {"type": "string"},
                        "description": "Validation command ids to require",
                    },
                    "artifact_overrides": {
                        "type": "object",
                        "description": (
                            "Name deliverables per pack role, e.g. "
                            '{"architecture_document": '
                            '{"dest_path": "docs/integration_testing_architecture.md"}}. '
                            "A dest_path alone also renames the produced artifact."
                        ),
                        "additionalProperties": {
                            "type": "object",
                            "properties": {
                                "logical_name": {"type": "string"},
                                "dest_path": {"type": "string"},
                            },
                            "additionalProperties": False,
                        },
                    },
                    "mock": {
                        "type": "boolean",
                        "description": "Force mock gateway",
                        "default": False,
                    },
                    "profile": {
                        "type": "string",
                        "description": "Model profile set",
                        "default": "local-target",
                    },
                },
                "required": ["request_text"],
            },
        },
        {
            "name": "pf_status",
            "description": "Get HostResponse status for a run_id.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "run_id": {"type": "string"},
                },
                "required": ["run_id"],
            },
        },
        {
            "name": "pf_tail",
            "description": (
                "Fetch one event batch after after_seq (non-blocking). "
                "Poll until terminal or awaiting_approval."
            ),
            "inputSchema": {
                "type": "object",
                "properties": {
                    "run_id": {"type": "string"},
                    "after_seq": {
                        "type": "integer",
                        "description": "Resume cursor (inclusive lower bound)",
                        "default": 0,
                    },
                },
                "required": ["run_id"],
            },
        },
        {
            "name": "pf_inspect",
            "description": "Inspect plan, validations, artifacts, and approval record.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "run_id": {"type": "string"},
                },
                "required": ["run_id"],
            },
        },
        {
            "name": "pf_approve",
            "description": "Approve a run awaiting operator approval.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "run_id": {"type": "string"},
                    "apply": {
                        "type": "boolean",
                        "description": "Apply patch to working tree when true",
                        "default": False,
                    },
                },
                "required": ["run_id"],
            },
        },
        {
            "name": "pf_reject",
            "description": "Reject a run awaiting operator approval.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "run_id": {"type": "string"},
                },
                "required": ["run_id"],
            },
        },
        {
            "name": "pf_cancel",
            "description": "Request cooperative cancel for a run.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "run_id": {"type": "string"},
                },
                "required": ["run_id"],
            },
        },
        {
            "name": "pf_export",
            "description": "Export a redaction-aware evidence bundle; path in data.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "run_id": {"type": "string"},
                    "as_zip": {
                        "type": "boolean",
                        "default": True,
                    },
                },
                "required": ["run_id"],
            },
        },
        {
            "name": "pf_materialize",
            "description": (
                "Copy a run artifact into the target repository under repository_path. "
                "Allowed when status is awaiting_approval or completed. "
                "Rejects path escape outside the repo root."
            ),
            "inputSchema": {
                "type": "object",
                "properties": {
                    "run_id": {"type": "string"},
                    "artifact": {
                        "type": "string",
                        "description": (
                            "Logical name (e.g. ARCHITECTURE.md, EVIDENCE_REPORT.md) "
                            "or artifact sha256"
                        ),
                    },
                    "dest_path": {
                        "type": "string",
                        "description": (
                            "Destination path relative to repository_path "
                            "(e.g. docs/ARCHITECTURE.md)"
                        ),
                    },
                    "overwrite": {
                        "type": "boolean",
                        "description": "Replace an existing destination file",
                        "default": False,
                    },
                },
                "required": ["run_id", "artifact", "dest_path"],
            },
        },
        {
            "name": "pf_materialize_all",
            "description": (
                "Land every deliverable of a run at its suggested destination from "
                "the run's artifact_land_map (see pf_inspect). Requires the same "
                "operator confirmation as pf_materialize; nothing lands implicitly."
            ),
            "inputSchema": {
                "type": "object",
                "properties": {
                    "run_id": {"type": "string"},
                    "roles": {
                        "type": "array",
                        "items": {"type": "string"},
                        "description": ("Limit to these deliverable roles (default: all landable)"),
                    },
                    "overwrite": {
                        "type": "boolean",
                        "description": "Replace existing destination files",
                        "default": False,
                    },
                },
                "required": ["run_id"],
            },
        },
    ]


def pf_submit(service: HostService, arguments: dict[str, Any]) -> dict[str, Any]:
    request_text = arguments.get("request_text")
    if not isinstance(request_text, str) or not request_text.strip():
        return _failure("invalid_arguments", "request_text is required")

    workflow = arguments.get("workflow") or "code_change"
    if workflow not in _WORKFLOW_VALUES:
        return _failure(
            "invalid_arguments",
            f"Unknown workflow {workflow!r}",
            details={"allowed": sorted(_WORKFLOW_VALUES)},
        )

    repo_raw = arguments.get("repository_path")
    repository_path: Path | None = None
    if repo_raw:
        repository_path = Path(str(repo_raw)).expanduser().resolve()

    budget_raw = arguments.get("budget_usd")
    budget_usd = 3.0 if budget_raw is None else float(budget_raw)
    validation_commands = arguments.get("validation_commands") or []
    if not isinstance(validation_commands, list):
        return _failure("invalid_arguments", "validation_commands must be a list of strings")
    mock = bool(arguments.get("mock", False))
    profile = str(arguments.get("profile") or "local-target")

    raw_overrides = arguments.get("artifact_overrides") or {}
    try:
        artifact_overrides = {
            role: ArtifactOverride.model_validate(spec)
            for role, spec in normalize_overrides(raw_overrides).items()
        }
    except (ArtifactOverrideError, ValueError) as exc:
        return _failure("invalid_artifact_override", str(exc))

    run_request = RunRequest(
        request_id=f"req-{uuid.uuid4().hex[:8]}",
        workflow_type=workflow,  # type: ignore[arg-type]
        request_text=request_text,
        repository_path=repository_path,
        model_profile_set=profile,
        validation_commands=[str(c) for c in validation_commands],
        artifact_overrides=artifact_overrides,
        budget=RunBudget(max_cost_usd=Decimal(str(budget_usd))),
    )
    return _as_json(
        service.submit(
            run_request,
            mock=mock,
            detach=True,
            inline_thread=False,
        )
    )


def pf_status(service: HostService, arguments: dict[str, Any]) -> dict[str, Any]:
    run_id = arguments.get("run_id")
    if not isinstance(run_id, str) or not run_id:
        return _failure("invalid_arguments", "run_id is required")
    return _as_json(service.status(run_id))


def pf_tail(service: HostService, arguments: dict[str, Any]) -> dict[str, Any]:
    run_id = arguments.get("run_id")
    if not isinstance(run_id, str) or not run_id:
        return _failure("invalid_arguments", "run_id is required")
    after_seq = int(arguments.get("after_seq") or 0)
    batch = next(
        service.tail(
            run_id,
            after_seq=after_seq,
            follow=False,
            max_idle_polls=1,
            stop_when_terminal=True,
        ),
        None,
    )
    if batch is None:
        return _failure("not_found", f"No events for {run_id}", run_id=run_id)
    return _as_json(batch)


def pf_inspect(service: HostService, arguments: dict[str, Any]) -> dict[str, Any]:
    run_id = arguments.get("run_id")
    if not isinstance(run_id, str) or not run_id:
        return _failure("invalid_arguments", "run_id is required")
    return _as_json(service.inspect(run_id))


def pf_approve(service: HostService, arguments: dict[str, Any]) -> dict[str, Any]:
    run_id = arguments.get("run_id")
    if not isinstance(run_id, str) or not run_id:
        return _failure("invalid_arguments", "run_id is required")
    apply = bool(arguments.get("apply", False))
    return _as_json(service.approve(run_id, apply=apply))


def pf_reject(service: HostService, arguments: dict[str, Any]) -> dict[str, Any]:
    run_id = arguments.get("run_id")
    if not isinstance(run_id, str) or not run_id:
        return _failure("invalid_arguments", "run_id is required")
    return _as_json(service.reject(run_id))


def pf_cancel(service: HostService, arguments: dict[str, Any]) -> dict[str, Any]:
    run_id = arguments.get("run_id")
    if not isinstance(run_id, str) or not run_id:
        return _failure("invalid_arguments", "run_id is required")
    return _as_json(service.cancel(run_id))


def pf_export(service: HostService, arguments: dict[str, Any]) -> dict[str, Any]:
    run_id = arguments.get("run_id")
    if not isinstance(run_id, str) or not run_id:
        return _failure("invalid_arguments", "run_id is required")
    as_zip = bool(arguments.get("as_zip", True))
    return _as_json(service.export_bundle(run_id, as_zip=as_zip))


def pf_materialize(service: HostService, arguments: dict[str, Any]) -> dict[str, Any]:
    run_id = arguments.get("run_id")
    if not isinstance(run_id, str) or not run_id:
        return _failure("invalid_arguments", "run_id is required")
    artifact = arguments.get("artifact")
    if not isinstance(artifact, str) or not artifact.strip():
        return _failure("invalid_arguments", "artifact is required")
    dest_path = arguments.get("dest_path")
    if not isinstance(dest_path, str) or not dest_path.strip():
        return _failure("invalid_arguments", "dest_path is required")
    overwrite = bool(arguments.get("overwrite", False))
    return _as_json(
        service.materialize(
            run_id,
            artifact=artifact.strip(),
            dest_path=dest_path.strip(),
            overwrite=overwrite,
        )
    )


def pf_materialize_all(service: HostService, arguments: dict[str, Any]) -> dict[str, Any]:
    run_id = arguments.get("run_id")
    if not isinstance(run_id, str) or not run_id:
        return _failure("invalid_arguments", "run_id is required")
    roles = arguments.get("roles") or []
    if not isinstance(roles, list):
        return _failure("invalid_arguments", "roles must be a list of strings")
    overwrite = bool(arguments.get("overwrite", False))
    return _as_json(
        service.materialize_all(
            run_id,
            roles=[str(role) for role in roles] or None,
            overwrite=overwrite,
        )
    )


_HANDLERS = {
    "pf_submit": pf_submit,
    "pf_status": pf_status,
    "pf_tail": pf_tail,
    "pf_inspect": pf_inspect,
    "pf_approve": pf_approve,
    "pf_reject": pf_reject,
    "pf_cancel": pf_cancel,
    "pf_export": pf_export,
    "pf_materialize": pf_materialize,
    "pf_materialize_all": pf_materialize_all,
}


def dispatch_tool(
    service: HostService,
    name: str,
    arguments: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Dispatch a tool by name; always returns HostResponse-shaped JSON."""
    handler = _HANDLERS.get(name)
    if handler is None:
        return _failure("unknown_tool", f"Unknown tool {name!r}")
    return handler(service, arguments or {})

File: integrations/opencode-plugin/src/tools.ts
/**
 * Product Factory plugin tools (model-facing).
 *
 * The tool *logic* lives in plain, dependency-free handlers so it can be unit
 * tested with a mock {@link PfClient} and a mock `ask`. `createPfTools` wraps
 * those handlers with OpenCode's `tool()` helper (injected by `index.ts`), so
 * this module never has to import `@opencode-ai/plugin` at load time.
 */

import type {
  ArtifactOverride,
  LandMapEntry,
  MaterializeInput,
  PfClient,
  PfResponse,
} from "./pf-client.js";
import { landMapFrom } from "./pf-client.js";

/** Workflows that end in a proposed patch → land via `approve(apply=true)`. */
export const PATCH_WORKFLOWS = new Set(["code_change", "repository_change"]);

/**
 * Last-resort artifact + destination per workflow.
 *
 * Only used when a run's `artifact_land_map` is unavailable (e.g. a host CLI
 * older than P4.A). The land map from `pf_inspect` is always preferred, so a run
 * that asked for `integration_testing_architecture.md` lands under that name.
 */
export const MATERIALIZE_DEFAULTS: Record<string, { artifact: string; dest: string }> = {
  technical_plan: { artifact: "ARCHITECTURE.md", dest: "docs/ARCHITECTURE.md" },
  architecture: { artifact: "ARCHITECTURE.md", dest: "docs/ARCHITECTURE.md" },
  repository_investigation: { artifact: "EVIDENCE_REPORT.md", dest: "docs/EVIDENCE_REPORT.md" },
  // A quality run has three deliverables; without a land map only the primary
  // report can be placed, so `pf_merge` lands that and reports what it skipped.
  quality_gate: { artifact: "QUALITY_FINDINGS.md", dest: "docs/QUALITY_FINDINGS.md" },
};

const AWAITING = "awaiting_approval";
const TERMINAL = new Set([
  "completed",
  "failed",
  "blocked",
  "budget_exhausted",
  "plan_rejected",
  "cancelled",
]);

/**
 * Minimal subset of OpenCode's `ToolContext` the handlers rely on. `ask`
 * requests operator permission and rejects when the user declines.
 */
export interface PfToolContext {
  directory?: string;
  worktree?: string;
  ask?: (input: {
    permission: string;
    patterns?: string[];
    always?: string[];
    metadata?: Record<string, unknown>;
  }) => Promise<void>;
}

export interface PfToolDeps {
  client: PfClient;
  /** Default budget for `pf_run` when the caller omits one. */
  defaultBudgetUsd?: number;
  /** Bounds for `pf_wait` polling. */
  wait?: { maxPolls?: number; intervalMs?: number; sleep?: (ms: number) => Promise<void> };
}

// ---------------------------------------------------------------------------
// Arg shapes (mirrors of the Zod schemas built in createPfTools)
// ---------------------------------------------------------------------------

export interface PfRunArgs {
  request: string;
  workflow?: string;
  repository_path?: string;
  budget_usd?: number;
  validation_commands?: string[];
  /**
   * Deliverable naming per pack role, e.g.
   * `{"architecture_document": "docs/integration_testing_architecture.md"}`.
   * A bare string is treated as the destination path.
   */
  artifact_overrides?: Record<string, string | ArtifactOverride>;
}

export interface PfWaitArgs {
  run_id: string;
  max_polls?: number;
}

export interface PfReviewArgs {
  run_id: string;
}

export interface PfMergeArgs {
  run_id: string;
  /** Override the detected workflow (patch vs doc/report). */
  workflow?: string;
  /** Artifact logical name; overrides the land map for a single-file merge. */
  artifact?: string;
  /** Destination path under the run repository; overrides the land map. */
  dest_path?: string;
  /** Limit a multi-deliverable merge to these land-map roles. */
  roles?: string[];
  overwrite?: boolean;
}

export interface PfDeclineArgs {
  run_id: string;
  /** `reject` (default for awaiting_approval) or `cancel` (in-flight run). */
  action?: "reject" | "cancel";
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function summarize(res: PfResponse): string {
  return JSON.stringify(
    {
      ok: res.ok,
      run_id: res.run_id,
      status: res.status,
      error: res.error ?? undefined,
      plan_summary: res.plan_summary ?? undefined,
      artifacts: res.artifacts?.map((a) => a.logical_name).filter(Boolean),
      data: res.data ?? undefined,
    },
    null,
    2,
  );
}

function workflowOf(res: PfResponse, override?: string): string | undefined {
  if (override) return override;
  const fromData = res.data?.["workflow_type"];
  return typeof fromData === "string" ? fromData : undefined;
}

const defaultSleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

// ---------------------------------------------------------------------------
// Handlers (pure — unit tested directly)
// ---------------------------------------------------------------------------

/** Accept `{role: "docs/x.md"}` as well as `{role: {destPath: "docs/x.md"}}`. */
export function normalizeArtifactOverrides(
  raw: Record<string, string | ArtifactOverride> | undefined,
): Record<string, ArtifactOverride> | undefined {
  if (!raw) return undefined;
  const out: Record<string, ArtifactOverride> = {};
  for (const [role, value] of Object.entries(raw)) {
    if (typeof value === "string") {
      if (value.trim()) out[role] = { destPath: value.trim() };
      continue;
    }
    const override: ArtifactOverride = {};
    if (value?.destPath?.trim()) override.destPath = value.destPath.trim();
    if (value?.logicalName?.trim()) override.logicalName = value.logicalName.trim();
    if (override.destPath || override.logicalName) out[role] = override;
  }
  return Object.keys(out).length > 0 ? out : undefined;
}

export async function pfRun(deps: PfToolDeps, ctx: PfToolContext, args: PfRunArgs): Promise<string> {
  const repositoryPath = args.repository_path ?? ctx.worktree ?? ctx.directory;
  const res = await deps.client.submit({
    requestText: args.request,
    workflow: args.workflow ?? "code_change",
    repositoryPath,
    budgetUsd: args.budget_usd ?? deps.defaultBudgetUsd,
    validationCommands: args.validation_commands,
    artifactOverrides: normalizeArtifactOverrides(args.artifact_overrides),
  });
  return summarize(res);
}

export async function pfWait(deps: PfToolDeps, _ctx: PfToolContext, args: PfWaitArgs): Promise<string> {
  const maxPolls = args.max_polls ?? deps.wait?.maxPolls ?? 60;
  const intervalMs = deps.wait?.intervalMs ?? 2_000;
  const sleep = deps.wait?.sleep ?? defaultSleep;
  let last: PfResponse | undefined;
  for (let poll = 0; poll < maxPolls; poll += 1) {
    last = await deps.client.status(args.run_id);
    const status = last.status ?? undefined;
    if (!last.ok) return summarize(last);
    if (status === AWAITING || (status && TERMINAL.has(status))) {
      return summarize(last);
    }
    if (poll < maxPolls - 1) await sleep(intervalMs);
  }
  return summarize({
    ...(last ?? { protocol: "product-factory.host/v1", ok: true, artifacts: [], events: [] }),
    data: { ...(last?.data ?? {}), timed_out: true, polls: maxPolls },
  } as PfResponse);
}

export async function pfReview(deps: PfToolDeps, _ctx: PfToolContext, args: PfReviewArgs): Promise<string> {
  const res = await deps.client.inspect(args.run_id);
  return summarize(res);
}

/**
 * Land a run's results into the workspace.
 *
 * INVARIANT: no `approve`/`materialize` call happens without an explicit
 * operator confirmation via `context.ask`. If `ask` is unavailable the merge is
 * refused (fail-closed) rather than silently proceeding.
 */
export async function pfMerge(deps: PfToolDeps, ctx: PfToolContext, args: PfMergeArgs): Promise<string> {
  const status = await deps.client.status(args.run_id);
  if (!status.ok) return summarize(status);

  const workflow = workflowOf(status, args.workflow);
  const isPatch = workflow ? PATCH_WORKFLOWS.has(workflow) : false;

  // A single explicit artifact/dest bypasses the land map; otherwise the run's
  // own land map decides which files land where.
  const singleFile = Boolean(args.artifact || args.dest_path);
  let landMap: LandMapEntry[] = [];
  if (!isPatch && !singleFile) {
    const inspected = await deps.client.inspect(args.run_id);
    if (inspected.ok) {
      landMap = landMapFrom(inspected).filter(
        (entry) => entry.landable !== false && (!args.roles || args.roles.includes(entry.role)),
      );
    }
  }

  const confirmed = await confirmMerge(ctx, {
    runId: args.run_id,
    workflow,
    action: isPatch
      ? "approve+apply patch"
      : landMap.length > 0
        ? `approve + land ${landMap.map((e) => e.suggested_dest_path).join(", ")}`
        : "approve + materialize document",
    destinations: landMap.map((entry) => entry.suggested_dest_path),
  });
  if (!confirmed) {
    return summarize({
      protocol: status.protocol,
      ok: false,
      run_id: args.run_id,
      status: status.status,
      artifacts: [],
      events: [],
      error: {
        code: "merge_declined",
        message: "Operator declined the merge; no approve/materialize was performed.",
      },
    });
  }

  if (isPatch) {
    const approved = await deps.client.approve(args.run_id, { apply: true });
    return summarize(approved);
  }

  // Doc/report workflow: approve if still awaiting, then land.
  if (status.status === AWAITING) {
    const approved = await deps.client.approve(args.run_id, { apply: false });
    if (!approved.ok) return summarize(approved);
  }

  // One confirmation covers the whole land map; the host still audits per file.
  if (landMap.length > 0) {
    const landed = await deps.client.materializeAll(args.run_id, {
      roles: args.roles,
      overwrite: args.overwrite ?? false,
    });
    return summarize(landed);
  }

  const preset = workflow ? MATERIALIZE_DEFAULTS[workflow] : undefined;
  const artifact = args.artifact ?? preset?.artifact ?? "ARCHITECTURE.md";
  const destPath = args.dest_path ?? preset?.dest ?? "docs/ARCHITECTURE.md";
  const input: MaterializeInput = { artifact, destPath, overwrite: args.overwrite ?? false };
  const materialized = await deps.client.materialize(args.run_id, input);
  return summarize(materialized);
}

export async function pfDecline(deps: PfToolDeps, _ctx: PfToolContext, args: PfDeclineArgs): Promise<string> {
  let action = args.action;
  if (!action) {
    const status = await deps.client.status(args.run_id);
    action = status.status === AWAITING ? "reject" : "cancel";
  }
  const res = action === "reject" ? await deps.client.reject(args.run_id) : await deps.client.cancel(args.run_id);
  return summarize(res);
}

async function confirmMerge(
  ctx: PfToolContext,
  info: { runId: string; workflow?: string; action: string; destinations?: string[] },
): Promise<boolean> {
  if (typeof ctx.ask !== "function") return false;
  try {
    await ctx.ask({
      permission: "pf_merge",
      patterns: [info.runId, ...(info.destinations ?? [])],
      always: [],
      metadata: {
        run_id: info.runId,
        workflow: info.workflow,
        action: info.action,
        destinations: info.destinations,
      },
    });
    return true;
  } catch {
    return false;
  }
}

// ---------------------------------------------------------------------------
// Plugin tool factory (wraps handlers with the injected `tool` helper)
// ---------------------------------------------------------------------------

/** The `tool` helper exported by `@opencode-ai/plugin` (injected to keep this module dependency-free). */
export type ToolHelper = {
  (input: {
    description: string;
    args: Record<string, unknown>;
    execute: (args: any, context: PfToolContext) => Promise<string>;
  }): unknown;
  schema: {
    string: () => any;
    number: () => any;
    boolean: () => any;
    array: (inner: unknown) => any;
    enum: (values: string[]) => any;
    record: (key: unknown, value: unknown) => any;
  };
};

export function createPfTools(tool: ToolHelper, deps: PfToolDeps): Record<string, unknown> {
  const s = tool.schema;
  return {
    pf_run: tool({
      description:
        "Submit a curated Product Factory run (workflow pack + request text). " +
        "Defaults repository_path to the current OpenCode workspace. Returns a run_id; " +
        "does NOT approve or apply anything. Send curated request text only, never the full chat.",
      args: {
        request: s.string().describe("Curated request text for the run (not the full transcript)."),
        workflow: s
          .string()
          .optional()
          .describe(
            "Workflow pack id (code_change, repository_change, technical_plan, " +
              "repository_investigation, quality_gate).",
          ),
        repository_path: s.string().optional().describe("Target repo; defaults to the OpenCode workspace root."),
        budget_usd: s.number().optional().describe("Cost ceiling in USD."),
        validation_commands: s.array(s.string()).optional().describe("Validation command ids to run."),
        artifact_overrides: s
          .record(s.string(), s.string())
          .optional()
          .describe(
            "Name the deliverables by pack role, e.g. " +
              '{"architecture_document": "docs/integration_testing_architecture.md"}. ' +
              "Use this whenever the user asks for a scoped document instead of the generic default.",
          ),
      },
      execute: (args: PfRunArgs, context: PfToolContext) => pfRun(deps, context, args),
    }),
    pf_wait: tool({
      description:
        "Poll a run until it reaches awaiting_approval or a terminal status (bounded polling). Returns the latest status envelope.",
      args: {
        run_id: s.string().describe("Run id returned by pf_run."),
        max_polls: s.number().optional().describe("Maximum status polls before returning."),
      },
      execute: (args: PfWaitArgs, context: PfToolContext) => pfWait(deps, context, args),
    }),
    pf_review: tool({
      description: "Inspect a run and summarize its plan, validations, and artifacts for the user before any merge.",
      args: { run_id: s.string().describe("Run id to inspect.") },
      execute: (args: PfReviewArgs, context: PfToolContext) => pfReview(deps, context, args),
    }),
    pf_merge: tool({
      description:
        "Land a run's results into the workspace. ALWAYS asks the operator for confirmation first. " +
        "Patch workflows are approved+applied; doc/report workflows are approved (if needed) and every " +
        "deliverable in the run's artifact_land_map is landed at its suggested path (see pf_review). " +
        "Pass artifact/dest_path only to override the land map for a single file.",
      args: {
        run_id: s.string().describe("Run id to merge."),
        workflow: s.string().optional().describe("Override detected workflow (patch vs doc/report)."),
        artifact: s.string().optional().describe("Artifact logical name for a single-file merge."),
        dest_path: s.string().optional().describe("Destination path for a single-file merge."),
        roles: s
          .array(s.string())
          .optional()
          .describe("Limit the merge to these land-map roles (default: every landable deliverable)."),
        overwrite: s.boolean().optional().describe("Overwrite existing destination files."),
      },
      execute: (args: PfMergeArgs, context: PfToolContext) => pfMerge(deps, context, args),
    }),
    pf_decline: tool({
      description:
        "Decline a run: reject (when awaiting_approval) or cancel (when still in flight). No workspace changes are made.",
      args: {
        run_id: s.string().describe("Run id to decline."),
        action: s.enum(["reject", "cancel"]).optional().describe("Force reject or cancel; auto-detected otherwise."),
      },
      execute: (args: PfDeclineArgs, context: PfToolContext) => pfDecline(deps, context, args),
    }),
  };
}
